src/MarginTrading.NotificationGenerator.Services/TradingReportService.cs

```diff
@@ -142,8 +142,149 @@ public async Task PerformReporting()
+
+                    await _log.WriteInfoAsync(nameof(NotificationGenerator), nameof(TradingReportService),
+                        nameof(PerformReporting), notification.GetLogData(), _systemClock.UtcNow.DateTime);
+                }
+                catch (Exception ex)
+                {
+                    anyException = ex;
+                    failedNotifications.Add(notification, ex);
+                }
+            }
+
+            if (failedNotifications.Any())
+            {
+                //TODO handle failed notifications
+                await _log.WriteWarningAsync(nameof(NotificationGenerator), nameof(TradingReportService),
+                    nameof(PerformReporting), $"{failedNotifications.Count} notifications failed to be send.",
+                    anyException, _systemClock.UtcNow.DateTime);
+            }
+        }
+
+        private static DailyTradingNotification PrepareNotification(int reportDay, 
+            DateTime from, DateTime to, string clientId, 
+            IReadOnlyList<OrderHistory> closedTrades, IReadOnlyList<OrderHistory> openPositions, 
+            IReadOnlyList<OrderHistory> pendingPositions, IReadOnlyList<Account> accounts,
+            IReadOnlyList<AccountHistory> accountTransactions)
+        {
+            var filteredAccounts = accounts.Where(x => x.ClientId == clientId)
+                .Select(x =>
+                {
+                    x.AccountTransactions = accountTransactions.Where(at => at.ClientId == clientId
+                                                                            && at.AccountId == x.Id)
+                        .OrderByDescending(at => at.Date).ToList();
+                    return x;
+                })
+                .OrderByDescending(x => x.Balance).ThenBy(x => x.BaseAssetId).ToList();
+            var accountIds = Enumerable.ToHashSet(filteredAccounts.Select(x => x.Id));
+            return new DailyTradingNotification
+            {
+                CurrentDay = from.ToString("dd.MM.yyyy"),
+                From = from.ToString("dd.MM.yyyy"),
+                To = to.ToString("dd.MM.yyyy"),
+                ClientId = clientId,
+                ClosedTrades = closedTrades
+                    .Where(x => x.ClientId == clientId && accountIds.Contains(x.AccountId))
+                    .OrderByDescending(x => x.CloseDate).ToList(),
+                OpenPositions = openPositions
+                    .Where(x => x.ClientId == clientId && accountIds.Contains(x.AccountId))
+                    .OrderByDescending(x => x.OpenDate).ToList(),
+                PendingPositions = pendingPositions
+                    .Where(x => x.ClientId == clientId && accountIds.Contains(x.AccountId))
+                    .OrderByDescending(x => x.CreateDate).ToList(),
+                Accounts = filteredAccounts,
+            };
+        }
+
+        private async Task SendNotificationsMonthly(IEnumerable<MonthlyTradingNotification> notifications,
```

Comment:
SendNotifications and SendNotificationsMonthly logic is quite the same, except time period labels. Maybe they can be merged together?
PerformReporting and PerformReportingMonthly are quite the same too. Let's extract the same part into private method and reuse it.

Reply:
I agree, but with further testing and review  logic has been changed due to the fact that we need to send only emails to people with open position and now the methods are not the same anymore

Reply:
Ok, thats fine then.